### PR TITLE
ENH: Fix CNT annotations to include bad spans

### LINKED ISSUE #11630: Importing bad data spans from Raw neuroscan cnt files into annotation description.
### Describe the new feature or enhancement

Hello! My lab has a lot of data that was processed using neuroscan, and I am trying to import some of our raw cnt files into mne. I have run into a problem with importing event files as neuroscan uses the same stim code to mark the end and the beginning of bad blocks (data spans): 
image
As far as I can tell, mne only imports the event code and the onset of these events (not the duration). However, I was looking through an old manual on neuroscan and it references another parameter that describes whether you should accept or reject a block:
image
I was wondering if this could be implemented by appending '_bad' to the end of the blocks that are meant to be rejected and manually computing the duration of the event by finding the difference in onsets of the event codes.

Another suggestion I have is that the  _translating_function does not convert the early event byte onsets properly to seconds as it says they are outside the data range (I used it myself and got a negative time). I suggest automatically setting event onset values to 0 if they have a cnt byte onset below 3665 (calculated using the translation function).

### Describe your proposed implementation

From cnt.py:
  ```
  with open(fname, 'rb') as fid:
        fid.seek(event_table_pos + 9)  # the real table stats at +9
        buffer = fid.read(teeg.total_length)

    my_events = list(event_parser(buffer))

    if not my_events:
        return Annotations(list(), list(), list(), None)
    else:
        onset = _translating_function(np.array([e.Offset for e in my_events],
                                               dtype=float),
                                      n_channels=n_channels,
                                      event_type=type(my_events[0]),
                                      data_format=data_format)
        duration = np.array([getattr(e, 'Latency', 0.) for e in my_events],
                            dtype=float)

        description = np.array([str(e.StimType) for e in my_events])
        return Annotations(onset=onset / sfreq,
                           duration=duration,
                           description=description,
                           orig_time=None)

```

- Make a new variable:  
`accept = np.array([str(e.KeyPad_Accept) for e in my_events])`
If it contains 'xc0' append '_bad' to the event code (description of event in mne)
Manually compute durations (currently all durations are 0) by finding the difference between codes.

- Set onset to 0 if .cnt byte onset is below 3664.9999999971237.

### Describe possible alternatives

I cannot think of any. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by withmywoessner**
#### Reference issue
Fixes #11630.

#### What does this implement/fix?
When importing raw neuroscan cnt files important information regarding bad spans is not included. This is because Neuroscan marks the start and end of bad blocks with event code 0. Currently, mne.io.read_raw_cnt only reads the description (StimType ), onset (Offset), and duration (Latency) of the events. However, there is another parameter KeyPad_Accept that includes whether the event code marks the beginning of a bad or good span (or neither). 
image

In my proposed changes I append 'bad_' to the codes that mark bad spans, calculate the difference in onsets between bad and good event codes to add to the duration parameter, and then drop the codes that mark a good block because once you have the duration of a bad block they are no longer needed.

Another suggestion I have is that the _translating_function does not convert the early event byte onsets properly to seconds as it says they are outside the data range (I used it myself and got a negative time) so I change a specific onset to 0 if the value in event_time is less than 0.

New helper functions:
https://github.com/withmywoessner/mne-python/blob/d93ad9256c47ffe2a063430b0fb5d2cad8e7b386/mne/io/cnt/cnt.py#L49-L108

Add new variable accept_reject:
https://github.com/withmywoessner/mne-python/blob/d93ad9256c47ffe2a063430b0fb5d2cad8e7b386/mne/io/cnt/cnt.py#L143-L153

Here is an example CNT to test my proposed changes (I don't know how to upload the file since it is more than 25mb so here is a Box download link):
https://wustl.box.com/s/lfhwycg8vnqvv3jgqf2gmfetpuce5hks
#### Additional information
Just to let everyone know here I am new to software development and Github so please let me know if I can make my code better in any way or if I am going about this process wrong :)

### CODE REVIEW

File: mne/io/cnt/cnt.py
Comment (drammock) [inline mne/io/cnt/cnt.py:84]: if you intend this to be visible to the user, it should be passed to our logger rather than printed.  try `git grep "logger.info"` to see examples of how it's done elsewhere in the codebase. (I'm assuming these should be informational messages, not warnings or errors)

(If you don't want the user to see it and it's just for your use during development/debugging, then it should be removed before this PR is finished)

**Review by drammock — COMMENTED** [review]
Thanks for tackling this!  I had a quick look and noted a few things; let us know when it's ready for another set of eyes again.  Feel free to ask if any of my suggestions aren't clear or you're not sure how to carry them out

**Comment by withmywoessner** [comment]
I updated my code with the suggested changes. Thank you @drammock for the suggestions! I think I am ready for it to be looked at again. I also ran make flake and got a few issues with my lines being too long. Should I remove my list comprehensions?

File: mne/io/cnt/cnt.py
Comment (withmywoessner) [inline mne/io/cnt/cnt.py:59]: I created this function but its functionality could definitely be implemented inside of _update_bad_span_onset. Should I do that or keep it as a separate function?

**Comment by drammock** [comment]
> Should I remove my list comprehensions?

No need to remove, you can split them over 2 lines

**Comment by withmywoessner** [comment]
I think my code is ready to be looked at again. I made it flake8 complaint. Apologies for all the commits I made. I was a bit confused about the process at first.

**Comment by drammock** [comment]
sorry for the delay; I'll try to look again early next week. (or if @agramfort is doing weekend PR reviews, maybe he'll get to this one :)

**Review by agramfort — COMMENTED** [review]
code works for me as expected.

@withmywoessner can you make an even smaller dataset so we can use it for testing? about or less than 10MB?

Then we can add it to the mne-testing-data repo and use it for testing here.

thanks a lot @withmywoessner !

**Comment by withmywoessner** [comment]
Sure! @agramfort
Here is a file that is around 9.7MB:
https://wustl.box.com/s/f76ush01pv18ofag6rzblqqokwlk9er4

Let me know if I need to add anything else before the merge. Thanks!

**Comment by withmywoessner** [comment]
@larsoner I just deleted all of the channels except 2 and it is now ~.8MB. Is that a good size? Let me know if you need anything else!
Link:
https://wustl.box.com/s/f76ush01pv18ofag6rzblqqokwlk9er4

**Comment by larsoner** [comment]
@withmywoessner awesome! If you want to continue learning/following our standard procedures, they would be:

1. Add file to mne-testing-data: [readme](https://github.com/mne-tools/mne-testing-data/#add-or-change-files-in-the-repo-for-use-with-mne)
2. Add a test here that fails on `main` but passes on your PR (probably in `mne/io/cnt/tests/test_cnt.py`)

If you want to give a shot feel free, otherwise I'm happy to do it!

**Comment by withmywoessner** [comment]
@larsoner I submitted a pull request in [mne-testing-data](https://github.com/mne-tools/mne-testing-data). I also tried creating a test that just checks to see if 'bad' is in the annotations description. Maybe, I should do something more involved??

**Comment by larsoner** [comment]
https://github.com/mne-tools/mne-testing-data/releases/tag/0.146 released, you should be able to continue following the mne-testing-data instructions to update mne/datasets

**Comment by larsoner** [comment]
I'm done pushing, feel free to take back over @withmywoessner :)

**Comment by withmywoessner** [comment]
Thanks @larsoner! I think I'm pretty much done, although I am not sure I wrote a proper test.

**Comment by larsoner** [comment]
Also if you're done/ready for merge don't forget to take the PR out of draft mode

File: mne/io/cnt/tests/test_cnt.py
Comment (withmywoessner) [inline mne/io/cnt/tests/test_cnt.py:61]:
```suggestion
    assert 'bad' in temp
```

**Comment by larsoner** [comment]
Looks good -- I think we just need an entry in `doc/changes/latest.inc` using `:newcontrib:` and URL in `doc/changes/names.inc` then we should be good to go! Note I pushed a tiny commit to move an entry that was in the wrong spot, but we still need one for this change

**Review by larsoner — APPROVED** [review]
Just fixed a tiny bug with the change doc (must link to a publicly documented function, not a private one) and will mark for merge when green, thanks in advance @withmywoessner !

**Pull request merged**
